docs/multiple_image_support_design_proposal.md

```diff
@@ -0,0 +1,549 @@
+    issues.
+  - Dependency between images must be described, which covers the API alignment
+    issue.
+  - Ensure atomic update when both images must be updated in one go, and
+    firmware cannot be started before updating of both images has properly
+    finished.
+  - In case one of the images needs to be reverted, it might be necessary to
+    revert the other one as well due to API incompatibilities.
+  - Flash will be more fragmented, as two memory slots are needed for each
+    image.
+  - Whether to use shared scratch area between images or use dedicated scratch
+    area per image?
+  - Whether to support image revert at all?
+
+# Assumptions and restrictions
+  - The proposed process is designed to work with more than two images.
+  - MCUBoot does the updates and authentication of all image, and if all checks
+    pass, it boots the image from primary slot of the 0th position.
+  - Other image in the boot chain is started by previous image.
+  - This proposal extends the current TLV manifest format. IETF manifest format
+    was out of scope.
+  - Revert feature of MCUBoot is also supported for multiple images.
+
+# Terminology and slot layout
+  - Header: image header
+  - Payload: image itself
+  - Manifest: contains TLV entries(image hash, key hash, signature, etc.)
+  - Image trailer:
+    - Swap status: progress info during swapping
+    - Swap state:  `struct image_trailer`
+
+The order of the items described above also represents the memory layout of a
+single slot. There can be some gaps (which is padded) between the following
+elements:
+  - header - payload: vector table must be on aligned address
+  - manifest - image trailer: image is smaller then memory slot
+
+```
+  +----------------------+
+  | Header               |   <- struct image_header
+  +----------------------+
+  | Payload              |   <- image
+  +----------------------+
+  | Manifest             |   <- TLV entries
+  +----------------------+
+  | Unused space in slot |   <- Filled with 0xFF
+  +----------------------+
+  | Trailer              |
+  | +------------------+ |
+  | | Swap status      | |   <- Progress info
+  | +------------------+ |
+  | | Swap state       | |   <- struct image_trailer
+  | +------------------+ |
+  +----------------------+
+```
+# Describing dependency
+As David Brown suggests the dependency between images can be described with
+additional TLV entries in the manifest.
+``` c
+#define IMAGE_TLV_DEPENDENCY  0x40   /* image depends on other image */
+
+#define IMAGE_CNT 2 /* Platform dependent, how many individual images are there
+                     */
+struct image_version {
+   uint8_t iv_major;
+   uint8_t iv_minor;
+   uint16_t iv_revision;
+   uint32_t iv_build_num;
+};
+
+struct image_dependency {
+    uint32_t image_id;  /* For simplicity could be just numbering from 0 to
+                         * IMAGE_CNT
+                         */
+    struct image_version image_min_version;  /* Indicates at minimum which
```

Comment:
What about image revert case, when one image (reverted one) require version form a range (probably exact version).
For simplicity it is possibile be satisfied by adding dependency on both depended image.

But in case when one image is provided external (so we can't modify it's) dependency TLV. And The second image has to keep exact range dependency.

Of course exact SHA works here pretty well, similar as addition of image_max_version field.

Reply:
Multi vendor scenario:
- Let's consider this scenario: There is an external vendor who provides the S binary, and there is the device owner who produces the NS binary and has access to the device to update both firmware. 
- I see two possible scenario how can this be handled.
- Scenario 1: Vendor shares the firmware with device owner. Device owner can verify the image authenticity at reception with PKI. Then device owner prepares the received S image for upgrade, which means putting header on it, concatenating TLV and dependency entry, calculating new hash and signature with his own private key (which public part is deployed in the device). Then send it to the device to upgrade. Also NS image must be prepared similar way. Might be different key pair is used to sign the S and NS image.
- Scenario 2: The bootloader contains two public keys. One of them is the corresponding part of the vendor's private key and used to verify the raw S image. The other is the corresponding part of the device owner's private key and used to verify the overhead of the S image and the NS image. In this case, the vendor will only provide the raw S binary with a signature, but the device owner can not verify it at reception, because the corresponding public key is only programmed to the device. Then the device owner still needs to prepare the image to be MCUBoot compatible: put header, include signature as TLV, could add dependency entry. In this case this additional data must also be protected, so there must be a secondary signature which also covers these items.
- Might the vendor can prepare the MCUBoot compatible  S image in some degree, add header (take into account at signing), and provide his own signature.. But probably the vendor does not add the device owner's dependency entry to the image in advanced, because he provides the firmware to many parties who requires different version number in the dependency entry.

```
  +--------------------+
  | Header             |   <- struct image_header
  +--------------------+
  | Payload            |
  +--------------------+
  | Manifest           |
  | +----------------+ |
  | | Manifest header| |   <- struct image_tlv_info
A | +----------------+ |
  | | SHA256 hash    | |   <- hash of header and payload (struct image_tlv)
B | +----------------+ |
  | | Keyhash        | |   <- Hash of vendor public key
  | +----------------+ |
  | | Signature-A-B  | |   <- Signature generated by vendor
C | +----------------+ |
  | | Dependency     | |   <- Device owner dependency entry
D | +----------------+ |
  | | Keyhash        | |   <- Hash of device owner public key
  | +----------------+ |
  | | Signature-C-D  | |   <- Signature generated by device owner
  | +----------------+ |
  +--------------------+
```

Reply:
As I understand scenario 1 assumes that mcuboot know only device owner public key? Looks like scenario 2 provide better security level (as mcuboot can verify independently Secure an Non-Secure image), and it is more flexible because by using the same keys it is possible to bring it almost to the 1st scenario.

Reply:
cc @frkv @oyvindronningstad

Reply:
I think we can assume that vendors who want to sign their own package will also sign their dependencies.  They will want to sign the complete description of the state needed to run their software. In my head, as stated below, dependencies are _not_ mutual (at the least not for the multiple signers case), which I guess simplifies this problem.

Reply:
This method of adding multi-image support seems somewhat similiar to IETF SUIT handling of nested manifests. Any reason why IETF SUIT method isn't considered as an alternative to a home-spun serialization? I don't see that this will scale if you get into more complex means of supporting multi image updates. As examples of a case I would like to see supported is doing a pass-through update to an external chip (banked in one of the slots) IETF SUIT supports this with the concept "ComponentIdentifier"

Reply:
@frkv - I would be not so critic - I have got in my mind iterative approach to the multiple image support - we can leave of simpler solution at the beginning, we can always redevelop it further.
Also IETF SUIT is still under discussion.

Reply:
That sounds good. Hopefully the work on this can somehow identify shortcomings in the SUIT standard. Do you by chance have more details written down about the eligibility of SUIT, and why it was kept out of this iteration?

Reply:
@nvlsianpu Yes, scenario 1 assumes that only the device owner public key is present in the device. But there is another key pair shared between SW vendor and device owner which is used to verify the SW component at reception.

 @oyvindronningstad Dependencies are not always mutual, but can be a case when they are (API or ABI changes between images). Regrading what to willing to sign a SW vendor can be vary among vendors. I assumes that a vendor who has many customers does not want to prepare and sign his SW product to meet all his customers needs. If images are not dependent on each other or we can assume that interface between them is quite mature, then it simplifies a lot of thing. 

The main goal of dependency handling/checking to verify whether an upgrade can be  started, all required SW components is in place. Second use case is during an image revert (question that this feature is needed at all) identify which other images must be reverted as well due to dependency.

@frkv  Originally in MCUBoot the TLV fields are used to describe the image's metadata.  I think IETF SUIT as an alternative for this. I think there is a separate discussion whether how and which level to support IETF SUIT in MCUBoot(I am not invovled). But currently it is not there, therefore I built my design proposal to the top of an existing feature.
What do you mean by: "pass-through update to an external chip"?

I will take a look to IETF SUIT to be more familiar. 

Do you have any preference/opinion on regarding to support auto revert feature?